design/lvm/resize_workflow.md

```diff
@@ -0,0 +1,153 @@
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [User Stories](#user-stories)
+    - [Implementation Details](#implementation-details)
+        - [Controller Expansion](#controller-expansion)
+        - [Filesystem Expansion](#filesystem-expansion)
+    - [Steps to perform volume expansion](#steps-to-perform-volume-expansion)
+    - [High level Sequence Diagram](#high-level-sequence-diagram)
+    - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Drawbacks](#drawbacks)
+  - [Alternatives](#alternatives)
+
+## Summary
+
+This proposal charts out the design details to implement expansion workflow on LVM-LocalPV Volumes.
+
+## Motivation
+
+### Goals
+
+- As a user, I should be able to resize the volumes provisioned via LVM-LocalPV by updating the size
+  on PersistentVolumeClaim(PVC).
+
+### Non-Goals
+
+- Support for volume expansion when underlying volume has lvm snapshots.
+
+## Proposal
+
+### User Stories
+
+As an application developer, I should be able to resize volume on the fly(When application consuming volume).
+
+Steps To Be Performed By User:
+1. Update the PVC capacity using `kubectl edit pvc <pvc_name>`
+
+Note: StorageClass should allow expanding the volume(i.e allowVolumeExpansion field must set to true).
+
+### Implementation Details
+
+LVM volume expansion is a two-step process(controller expansion & filesystem expansion) and
+it gets triggered upon updating PVC capacity.
+
+##### Controller Expansion
+
+- External-resizer is a sidecar in `openebs-lvm-controller` pod which watches for PVC capacity
+  updates. Once it receives the event external-resizer will make a `ControllerExpandVolume` gRPC request
+  to LVM-CSI controller plugin.
+- LVM-CSI controller plugin is another container in the same pod where external-resizer is running.
+  Once the controller plugin receives the `ControllerExpandVolume` gRPC request it does the following steps:
+  - List active snapshots on resizing volume, if there are any snapshots plugin will return an error as
+    a response to the request since LVM does not support online resizing of volume if snapshot(s) exist.
+  - If there are no snapshots then the plugin will update the capacity on the corresponding LVMVolume
+    resource and return success response to the request.
+- Once external-resizer gets success response it will mark PVC pending for filesystem expansion else
+  it will retry until it receives success response.
+
+Note: `ControllerExpandVolume` gRPC request retrun an error if volume has snapshots.
+
+##### Filesystem Expansion
+
+- Container Orchestrator(CO)[kubelet] will send `NodeExpandVolume` gRPC request to LVM-CSI node driver
+  once it observes the PVC marked for filesystem expansion.
+- LVM-CSI node driver receives the `NodeExpandVolume` gRPC request and it will expand the lvm volume
+  using lvextend cli and trigger filesystem expansion. If expansion is successful it will return
+  success gRPC response.
```

Comment:
If filesystem-resize is errored, will it go into reconcile mode till it is successful?

Reply:
Yes